[TDD and TRACEBACKS] new docs for development  and troubleshooting (#3059)
exercism/python: docs/TRACEBACKS.md

Once function `C` returns, then its frame object will be popped off the stack and only the frame objects for functions `A` and `B` will remain on the call stack.

## Traceback

A Traceback is a report of all the frame objects on the stack for a specific time.
When a Python program encounters an unhandled exception, it will print the exception message and a Traceback.
The Traceback will show where the exception was raised and what functions were called leading up to it.

## How to Read a Traceback

`ValueError` is a common exception.

Following is an example of `ValueError` resulting from trying to assign two variables on the left from only one value on the right:

```python
>>> first, second = [1]
Traceback (most recent call last):
File <stdin>, line 1, in <module>
    first, second = [1]
ValueError: not enough values to unpack (expected 2, got 1)

```

It says that the most recent call is last, so the place to start is with the exception at the bottom of the trace.
From there it can be traced back to the start.
If we place the offending line in a function and then call the function we see a longer trace:

```python
>>> def my_func():
...     first, second = [1]

> ... 

Reviewer: ```suggestion
...
```

>>> my_func()
Traceback (most recent call last):
  File <stdin>, line 5, in <module>
    my_func()
  File <stdin>, line 2, in my_func
    first, second = [1]
ValueError: not enough values to unpack (expected 2, got 1)

```

Working backwards from the bottom, we see that the call where the exception happened is on line 2 in `my_func`.
We got there by calling `my_func` on line 5.

## Using the `print` function

Sometimes an error is not being raised, but a value is not what is expected.
This can be especially perplexing if the value is the result of a chain of calculations.
In such a situation it can be helpful to look at the value at each step to see which step is the one that isn't behaving as expected.
The [print][print] function can be used for printing the value to the console.

> Following is an example of a function that doesn't return the value expected

Reviewer: ```suggestion
The following is an example of a function that doesn't return the value expected:
```

"The " is optional as a change, but the colon is missing at the end of this line.

```python
# the intent is to only pass an int to this function and get an int back
def halve_and_quadruple(num):
    return (num / 2) * 4
```

When the function is passed `5`, the expected value is `8`, but it it returns `10.0`.
To troubleshoot, the calculating is broken up so that the value can be inspected at every step.

```python
# the intent is to only pass an int to this function and get an int back
def halve_and_quadruple(num):
    # verify the number in is what is expected
    # prints 5
    print(num)
    # we want the int divided by an int to be an int
    # but this prints 2.5! We've found our mistake.
    print(num / 2)
    # this makes sense, since 2.5 x 4 = 10.0
    print((num / 2) * 4)
    return (num / 2) * 4

```

What the `print` calls revealed is that we used `/` when we should have used `//`, the [floor divison operator][floor divison operator].

## Logging

[Logging][logging] can be used similarly to `print`, but it is more powerful.

> Logging can print to the console, but it can also be configured to print to a file.

Reviewer: But we _can_ `print` to a file, so this comparison and reason why logging is "more powerful" is not really an honest reasoning.

Can we maybe come up with a better reason why we should use the logging facility?  Perhaps rotation of logs is built in, or some other feature of logging that `print` does not consider doing.

Author: I'm probably a bit out of my depth here, but I'll try to work in something from [this](https://stackoverflow.com/questions/6918493/in-python-why-use-logging-instead-of-print).

Reviewer: Things like loggers generally flush on any output, to ensure that logging is done, unlike functions that output to the terminal.  The terminal may also have controls over buffer to standard out and standard error.  Logging facilities often, also, try to ensure no loss of data, so may have safeguards to write even after most "catastrophic" failures in the program runtime.

These need should be confirmed for the language, but would be reasons why I would suggest logging over the normal way of output in a language, often via `print`.

Author: Okay, I added  a bit to better and more accurately give reasons for logging.

Reviewer: Downloading the latest changes to review… Will check in about 30, as Isaac is busy here as well, recently.

Following is a an example of logging printed to the console:

```python

> >>> import logging
> ... 
> ... # configures minimum logging level as INFO
> ... logging.basicConfig(level=logging.INFO)
> 
> ... def halve_and_quadruple(num):
> ...     # prints INFO:root: num == 5
> ...     logging.info(f" num == {num}")
> ...     return (num // 2) * 4
> ... 

Reviewer: ```suggestion
>>> import logging
...
... # configures minimum logging level as INFO
... logging.basicConfig(level=logging.INFO)

... def halve_and_quadruple(num):
...     # prints INFO:root: num == 5
...     logging.info(f" num == {num}")
...     return (num // 2) * 4
...
```

>>> print(halve_and_quadruple(5))

```

The level is configured as `INFO` because the default level is `WARNING`.
For a persistent log, the logger can be configured to use a file, like so:

```python

> >>> import logging
> ... 
> ... # configures the output file name to example.log, and the minimum logging level as INFO
> ... logging.basicConfig(filename='example.log', level=logging.INFO)
> ... 
> ... def halve_and_quadruple(num):
> ...     # prints INFO:root: num == 5 to the example.log file
> ...     logging.info(f" num == {num}")
> ...     return (num // 2) * 4
> ... 

Reviewer: ```suggestion
>>> import logging
...
... # configures the output file name to example.log, and the minimum logging level as INFO
... logging.basicConfig(filename='example.log', level=logging.INFO)
...
... def halve_and_quadruple(num):
...     # prints INFO:root: num == 5 to the example.log file
...     logging.info(f" num == {num}")
...     return (num // 2) * 4
...
```

## Frame Object

When a function is called, a frame object is created to hold the local variables and arguments passed to a function.
When the function is returns, the frame object is destroyed.
When function `B` is called within function `A`, function `B`'s values are put into a frame object, which is then placed on top of function `A`'s frame object on the call stack.

## Call Stack

The call stack is a collection of frame objects for the currently active functions.
If function `A` has called function `B` and function `B` has called function `C`, then the frame objects for all three functions will be on the call stack.
Once function `C` returns, then its frame object will be popped off the stack and only the frame objects for functions `A` and `B` will remain on the call stack.

## Traceback

A Traceback is a report of all the frame objects on the stack at a specific time.
When a Python program encounters an unhandled exception, it will print the exception message and a Traceback.
The Traceback will show where the exception was raised and what functions were called leading up to it.

## How to Read a Traceback

`ValueError` is a common exception.

> Following is an example of `ValueError` resulting from trying to assign two variables on the left from only one value on the right:

Reviewer: ```suggestion
The following is an example of `ValueError` resulting from trying to assign two variables on the left from only one value on the right:
```

```

When the function is passed `5`, the expected value is `8`, but it it returns `10.0`.
To troubleshoot, the calculating is broken up so that the value can be inspected at every step.

```python
# the intent is to only pass an int to this function and get an int back
def halve_and_quadruple(num):
    # verify the number in is what is expected
    # prints 5
    print(num)
    # we want the int divided by an int to be an int
    # but this prints 2.5! We've found our mistake.
    print(num / 2)
    # this makes sense, since 2.5 x 4 = 10.0
    print((num / 2) * 4)
    return (num / 2) * 4

```

What the `print` calls revealed is that we used `/` when we should have used `//`, the [floor divison operator][floor divison operator].

## Logging

[Logging][logging] can be used similarly to `print`, but it is more powerful.
What is logged can be configured by the logging severity (e.g., 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'.)
A call to the `logging.error` function can pass `True` to the `exc_info` parameter, which will additionally log the stack trace.
By configuring multiple handlers, logging can write to more than one place with the same logging function.

> Following is a an example of logging printed to the console:

Reviewer: ```suggestion
Following is an example of logging printed to the console:
```

Use "a" or "an" but not both.

Missing the "The " at the beginning as well.

```python

>>> import logging
...
... # configures minimum logging level as INFO
... logging.basicConfig(level=logging.INFO)

... def halve_and_quadruple(num):
...     # prints INFO:root: num == 5
...     logging.info(f" num == {num}")
...     return (num // 2) * 4
...
>>> print(halve_and_quadruple(5))

```

The level is configured as `INFO` because the default level is `WARNING`.

> For a persistent log, the logger can be configured to use a file, like so:

Reviewer: Instead of "use a file" we can be precise and say "write to a file".

Author: Addressed in separate edit.

```python

>>> import logging
...
... # configures the output file name to example.log, and the minimum logging level as INFO
... logging.basicConfig(filename='example.log', level=logging.INFO)
...
... def halve_and_quadruple(num):
...     # prints INFO:root: num == 5 to the example.log file
...     logging.info(f" num == {num}")
...     return (num // 2) * 4
...
>>> print(halve_and_quadruple(5))

```

## assert

[`assert`][assert] is a statement which should always evaluate to `True` unless there is a bug in the program.
When an `assert` evaluates to `False` it will raise an [`AssertionError`][AssertionError].
The Traceback for the `AssertionError` can include an optional message that is part of the `assert` statement.
Although a message is optional, it is good practice to always include one in the `assert` definition.

> Following is an example of using `assert`:

Reviewer: ```suggestion
The following is an example of using `assert`:
```

```python
>>> def int_division(dividend, divisor):
...     assert divisor != 0, "divisor must not be 0"
...     return dividend // divisor

> ... 

Reviewer: ```suggestion
...
```

>>> print(int_division(2, 1))
2
>>> print(int_division(2, 0))
Traceback (most recent call last):
  File <stdin>, line 7, in <module>
    print(int_division(2, 0))
          ^^^^^^^^^^^^^^^^^^
  File <stdin>, line 2, in int_division
    assert divisor != 0, "divisor must not be 0"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: divisor must not be 0

```

If we start reading the Traceback at the bottom (as we should) we quickly see the problem is that `0` should not be passsed as the `divisor`.

`assert` can also be used to test that a value is of the expected type:

```python
>>> import numbers

> ... 
> ... 

Reviewer: ```suggestion
...
...
```

... def int_division(dividend, divisor):
...     assert divisor != 0, "divisor must not be 0"
...     assert isinstance(divisor, numbers.Number), "divisor must be a number"
...     return dividend // divisor

> ... 

Reviewer: ```suggestion
...
```

>>> print(int_division(2, 1))
2
>>> print(int_division(2, '0'))
Traceback (most recent call last):
  File <stdin>, line 11, in <module>
    print(int_division(2, '0'))
          ^^^^^^^^^^^^^^^^^^^^
  File <stdin>, line 6, in int_division
    assert isinstance(divisor, numbers.Number), "divisor must be a number"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: divisor must be a number
```

> Once a bug is identified, consider replacing the `assert` with regular error handling.

Reviewer: ```suggestion
Once a bug is identified, consider replacing the `assert` with error handling.
```

> This is because all `assert` statements can disabled through running Python with the `-O` or `-OO` options, or from setting the `PYTHONOPTIMIZE` environment variable to `1` or `2`.

Reviewer: ```suggestion
This is because all `assert` statements can be disabled through running Python with the `-O` or `-OO` options, or from setting the `PYTHONOPTIMIZE` environment variable to `1` or `2`.
```

# Reading and Troubleshooting Stack Traces in Python

## Frame Object

> A frame object holds the local variables and arguments passed to a function.
> A frame object is created when a function is called and is destroyed when that function returns.

Reviewer 2: This starts with the new concept (frame) vs with the background/context (when a function is called).

```suggestion
When a function is called, a frame object is created to hold the local variables and arguments passed to a function.
When the function is returns, the frame object is destroyed.
```

> When function `B` is called within function `A`, function `B`'s values are put into a frame object, which is then placed on top of function `A`'s frame object on the call stack.

Reviewer 2: Should this be in the Call Stack section?

## Call Stack

The call stack is a collection of frame objects for the currently active functions.
If function `A` has called function `B` and function `B` has called function `C`, then the frame objects for all three functions will be on the call stack.
Once function `C` returns, then its frame object will be popped off the stack and only the frame objects for functions `A` and `B` will remain on the call stack.

## Traceback

> A Traceback is a report of all the frame objects on the stack for a specific time.

Reviewer 2: ```suggestion
A Traceback is a report of all the frame objects on the stack at a specific time.
```

When a Python program encounters an unhandled exception, it will print the exception message and a Traceback.
The Traceback will show where the exception was raised and what functions were called leading up to it.

## How to Read a Traceback

`ValueError` is a common exception.

Following is an example of `ValueError` resulting from trying to assign two variables on the left from only one value on the right:

```python
>>> first, second = [1]
Traceback (most recent call last):
File <stdin>, line 1, in <module>
    first, second = [1]

> ValueError: not enough values to unpack (expected 2, got 1)
> 

Reviewer 2: ```suggestion
ValueError: not enough values to unpack (expected 2, got 1)
```

```

> It says that the most recent call is last, so the place to start is with the exception at the bottom of the trace.
> From there it can be traced back to the start.

Reviewer 2: ```suggestion
Tracebacks are organized with the most recent call is last, so the place to start reading the Traceback is with the exception at the bottom of the output.
By reading upwards, we can see how that statement was reached.
```

Author: I like the upwards, but "Tracebacks are organized with the most recent call is last," isn't right .

Author: Addressed in separate edit.

If we place the offending line in a function and then call the function we see a longer trace:

```python
>>> def my_func():
...     first, second = [1]
...
>>> my_func()
Traceback (most recent call last):
  File <stdin>, line 5, in <module>
    my_func()
  File <stdin>, line 2, in my_func
    first, second = [1]

> ValueError: not enough values to unpack (expected 2, got 1)
> 

Reviewer 2: ```suggestion
ValueError: not enough values to unpack (expected 2, got 1)
```

```

Working backwards from the bottom, we see that the call where the exception happened is on line 2 in `my_func`.
We got there by calling `my_func` on line 5.

## Using the `print` function

Sometimes an error is not being raised, but a value is not what is expected.
This can be especially perplexing if the value is the result of a chain of calculations.
In such a situation it can be helpful to look at the value at each step to see which step is the one that isn't behaving as expected.
The [print][print] function can be used for printing the value to the console.
The following is an example of a function that doesn't return the value expected:

```python

> # the intent is to only pass an int to this function and get an int back

Reviewer 2: ```suggestion
# the intent is to pass an int to this function and get an int back
```

Author: Addressed in separate edit.

def halve_and_quadruple(num):
    return (num / 2) * 4
```

When the function is passed `5`, the expected value is `8`, but it it returns `10.0`.
To troubleshoot, the calculating is broken up so that the value can be inspected at every step.

> 

Reviewer: ```suggestion
```

```python
# the intent is to only pass an int to this function and get an int back
def halve_and_quadruple(num):
    # verify the number in is what is expected
    # prints 5
    print(num)
    # we want the int divided by an int to be an int
    # but this prints 2.5! We've found our mistake.
    print(num / 2)
    # this makes sense, since 2.5 x 4 = 10.0
    print((num / 2) * 4)
    return (num / 2) * 4

> 
> ```

Reviewer 2: ```suggestion
```
```

> What the `print` calls revealed is that we used `/` when we should have used `//`, the [floor divison operator][floor divison operator].

Reviewer 2: ```suggestion
The `print` calls revealed is that we used `/` when we should have used `//`, the [floor divison operator][floor divison operator].
```

## Logging

[Logging][logging] can be used similarly to `print`, but it is more powerful.
What is logged can be configured by the logging severity (e.g., 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'.)
A call to the `logging.error` function can pass `True` to the `exc_info` parameter, which will additionally log the stack trace.

> By configuring multiple handlers, logging can print to more than one place with the same logging function.

Reviewer 2: ```suggestion
Log entries are generated with a logging severity level (e.g., 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL').
Logging can be configured to only output log entries above a given severity.
The `logging.error` function accepts an optional `exc_info` parameter; if True, the stack trace will be added to the log entry.
By configuring multiple handlers, a logged entry can be written to multiple outputs or sinks.
```

Author: Leaving it.

> Following is a an example of logging printed to the console:

Reviewer 2: ```suggestion
Following is a an example of using logging to print log entries to the console:
```

> ```python
> 

Reviewer 2: ```suggestion
```python
```

>>> import logging

> ...
> ... # configures minimum logging level as INFO
> ... logging.basicConfig(level=logging.INFO)
> 
> ... def halve_and_quadruple(num):

Reviewer 2: ```suggestion
>>>
>>> # configures minimum logging level as INFO
>>> logging.basicConfig(level=logging.INFO)
>>>
>>> def halve_and_quadruple(num):
```

...     # prints INFO:root: num == 5

> ...     logging.info(f" num == {num}")

Reviewer 2: Some linters will complain that you should use format strings to allow for lazy formatting.

...     return (num // 2) * 4
...
>>> print(halve_and_quadruple(5))

```

> The level is configured as `INFO` because the default level is `WARNING`.

Reviewer 2: ```suggestion
The default log level is `WARNING`; in order for an INFO log to be generated, we set the level to `INFO` in the configuration.
```

> For a persistent log, the logger can be configured to use a file, like so:

Reviewer 2: ```suggestion
For a persistent log, the logger can be configured to use a file.
```

```python

> 

Reviewer 2: Drop the leading and trailing blank lines. `...` should be `>>>` on the first half of these lines.

Author: Thanks. Addressed in separate edit.

>>> import logging
...
... # configures the output file name to example.log, and the minimum logging level as INFO
... logging.basicConfig(filename='example.log', level=logging.INFO)
...
... def halve_and_quadruple(num):
...     # prints INFO:root: num == 5 to the example.log file
...     logging.info(f" num == {num}")
...     return (num // 2) * 4
...
>>> print(halve_and_quadruple(5))

```

## assert

> [`assert`][assert] is a statement which should always evaluate to `True` unless there is a bug in the program.

Reviewer 2: ```suggestion
[`assert`][assert] is a statement which allows us to assert that something is true; if it is not true, this indicated there is a bug in the program.
```

When an `assert` evaluates to `False` it will raise an [`AssertionError`][AssertionError].

> The Traceback for the `AssertionError` can include an optional message that is part of the `assert` definition.

Reviewer 2: ```suggestion
The Traceback for the `AssertionError` can include an optional message that is part of the `assert` statement.
```

  File <stdin>, line 7, in <module>
    print(int_division(2, 0))
          ^^^^^^^^^^^^^^^^^^
  File <stdin>, line 2, in int_division
    assert divisor != 0, "divisor must not be 0"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: divisor must not be 0

```

If we start reading the Traceback at the bottom (as we should) we quickly see the problem is that `0` should not be passsed as the `divisor`.

`assert` can also be used to test that a value is of the expected type:

```python
>>> import numbers
... 
... 
... def int_division(dividend, divisor):
...     assert divisor != 0, "divisor must not be 0"
...     assert isinstance(divisor, numbers.Number), "divisor must be a number"
...     return dividend // divisor
... 
>>> print(int_division(2, 1))
2
>>> print(int_division(2, '0'))
Traceback (most recent call last):
  File <stdin>, line 11, in <module>
    print(int_division(2, '0'))
          ^^^^^^^^^^^^^^^^^^^^
  File <stdin>, line 6, in int_division
    assert isinstance(divisor, numbers.Number), "divisor must be a number"
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: divisor must be a number
```

> Once a bug is identified, it should be considered to replace the `assert` with regular error handling.

Reviewer 2: ```suggestion
Once a bug is identified, consider replacing the `assert` with regular error handling.
```

> This is because all `assert` statements can disabled through running Python with the `-O` or `-OO` options, or from setting the `PYTHONOPTIMIZE` environment variable to `1` or `2`.
> Setting `PYTHONOPTIMIZE` to `1` is equivalent to running Python with the `-O` option, which disables assertions.
> Setting `PYTHONOPTIMIZE` to `2` is equivalent to running Python with the `-OO` option, which both disables assertions and removes docstrings from the bytcode.
> Reducing bytecode is one way to make the code run faster.

Reviewer 2: ```suggestion
This is because all optimizing Python code will disable all the `assert` statements.
Python can run in an optimized mode using `-O` or `-OO` options, or from setting the `PYTHONOPTIMIZE` environment variable to `1` or `2`.
Setting `PYTHONOPTIMIZE` to `1` is equivalent to running Python with the `-O` option, which disables assertions.
Setting `PYTHONOPTIMIZE` to `2` is equivalent to running Python with the `-OO` option, which both disables assertions and removes docstrings from the byetcode.
Both of these help make the code run faster.
```

Author: Another change addressed this

## Logging

[Logging][logging] can be used similarly to `print`, but it is more powerful.
What is logged can be configured by the logging severity (e.g., 'DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'.)
A call to the `logging.error` function can pass `True` to the `exc_info` parameter, which will additionally log the stack trace.

> By configuring multiple handlers, logging can print to more than one place with the same logging function.

Reviewer: Because of the closeness of the function with the word "print"` we might use "store or display" rather than "print".  "Output" might be another word, but that is often used even more "loosely".

Still, this is better.

Author: I see what you mean. Let me come up with a synonym...

Author: Maybe "persist"?

Reviewer: That data persists, the logger does not.  So I do not think that is the right word.  The logger can write to standard out, standard error, or a file.  And "write" it does.

Reviewer 2: Output? Write? Record? Logs aren't always printed (eg Datadog) nor persisted (mem based).
